typed/typed.go

```diff
@@ -187,7 +195,33 @@ func (tv TypedValue) RemoveItems(items *fieldpath.Set) *TypedValue {
 }
 
 // ExtractItems returns a value with only the provided list or map items extracted from the value.
-func (tv TypedValue) ExtractItems(items *fieldpath.Set) *TypedValue {
+func (tv TypedValue) ExtractItems(items *fieldpath.Set, opts ...ExtractItemsOptions) *TypedValue {
+	for _, opt := range opts {
+		switch opt {
+		case AppendKeyFields:
+			tvPathSet, err := tv.ToFieldSet()
+			if err != nil {
+				continue
+			}
+			keyFieldPathSet := fieldpath.NewSet()
+			items.Iterate(func(path fieldpath.Path) {
+				if !tvPathSet.Has(path) {
+					return
+				}
+				for i, pe := range path {
+					if pe.Key == nil {
+						continue
+					}
+					for _, keyField := range *pe.Key {
+						keyName := keyField.Name
+						keyFieldPath := append(path[:i+1:i+1], fieldpath.PathElement{FieldName: &keyName})
```

Comment:
We may have to make a copy of path, at least I've run into some issues when not doing so in my workaround.

Reply:
The `copy` doesn't make the difference here as the fields in the `PathElement` are pointers and they would still be copied to the new path, so that the new path still points to the same `PathElement`.

I think what matters is creating a new string `keyName := keyField.Name` in the for-loop, so that the appended `PathElement.FieldName` won't be overwritten on the same address.

Or, could you share the issues you encounter when not doing copy?

Reply:
Ah, maybe you are right. I have https://github.com/argoproj/gitops-engine/pull/640/commits/e881697fbaf4e3fc770eb7e0dd365606e4cffe8c, but I indeed didn't create a new string, so that might have been an issue.

Reply:
I'm fine with setting capacity here as a way to force an inline copy to be created.  I also wouldn't mind an extra comment for future readers stating what happens here since this is use of the language that not all readers may pick up on easily.

Reply:
I've added the comment.